File: squire-backend/api/routes/knowledge.js
// Knowledge Graph Routes
const express = require('express');
const { createClient } = require('@supabase/supabase-js');
const router = express.Router();

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_ANON_KEY
);

// POST /api/knowledge/nodes - Create/update knowledge node
router.post('/nodes', async (req, res) => {
  try {
    const {
      user_id,
      node_type,
      content,
      weight = 1.0,
      source_event_ids = [],
      metadata = {}
    } = req.body;

    const { data, error } = await supabase
      .rpc('upsert_knowledge_node', {
        p_user_id: user_id,
        p_node_type: node_type,
        p_content: content,
        p_weight: weight,
        p_source_event_ids: source_event_ids,
        p_metadata: metadata
      });

    if (error) throw error;

    // Get the created/updated node
    const { data: node, error: nodeError } = await supabase
      .from('knowledge_nodes')
      .select('*')
      .eq('id', data)
      .single();

    if (nodeError) throw nodeError;

    res.status(201).json({ node, node_id: data });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/knowledge/nodes/user/:userId - Get all nodes for user
router.get('/nodes/user/:userId', async (req, res) => {
  try {
    const { userId } = req.params;
    const {
      node_type,
      min_weight,
      limit = 100,
      offset = 0,
      order_by = 'weight',
      order_direction = 'desc'
    } = req.query;

    let query = supabase
      .from('knowledge_nodes')
      .select('*')
      .eq('user_id', userId);

    if (node_type) {
      query = query.eq('node_type', node_type);
    }

    if (min_weight) {
      query = query.gte('weight', parseFloat(min_weight));
    }

    query = query
      .order(order_by, { ascending: order_direction === 'asc' })
      .range(offset, offset + limit - 1);

    const { data, error } = await query;

    if (error) throw error;

    res.json({ nodes: data || [] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/knowledge/nodes/:id - Get specific node
router.get('/nodes/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const { data, error } = await supabase
      .from('knowledge_nodes')
      .select('*')
      .eq('id', id)
      .single();

    if (error) throw error;
    if (!data) return res.status(404).json({ error: 'Node not found' });

    res.json({ node: data });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// PUT /api/knowledge/nodes/:id - Update node
router.put('/nodes/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { content, weight, metadata } = req.body;

    const updateData = {};
    if (content !== undefined) updateData.content = content;
    if (weight !== undefined) updateData.weight = weight;
    if (metadata !== undefined) updateData.metadata = metadata;

    const { data, error } = await supabase
      .from('knowledge_nodes')
      .update(updateData)
      .eq('id', id)
      .select()
      .single();

    if (error) throw error;
    if (!data) return res.status(404).json({ error: 'Node not found' });

    res.json({ node: data });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// DELETE /api/knowledge/nodes/:id - Delete node
router.delete('/nodes/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const { data, error } = await supabase
      .from('knowledge_nodes')
      .delete()
      .eq('id', id)
      .select()
      .single();

    if (error) throw error;
    if (!data) return res.status(404).json({ error: 'Node not found' });

    res.json({ success: true, message: 'Node deleted successfully' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// POST /api/knowledge/relationships - Create/update relationship
router.post('/relationships', async (req, res) => {
  try {
    const {
      user_id,
      source_node_id,
      target_node_id,
      relationship_type,
      strength = 0.5,
      metadata = {}
    } = req.body;

    const { data, error } = await supabase
      .rpc('upsert_knowledge_relationship', {
        p_user_id: user_id,
        p_source_node_id: source_node_id,
        p_target_node_id: target_node_id,
        p_relationship_type: relationship_type,
        p_strength: strength,
        p_metadata: metadata
      });

    if (error) throw error;

    // Get the created/updated relationship
    const { data: relationship, error: relError } = await supabase
      .from('knowledge_relationships')
      .select('*')
      .eq('id', data)
      .single();

    if (relError) throw relError;

    res.status(201).json({ relationship, relationship_id: data });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/knowledge/relationships/user/:userId - Get all relationships for user
router.get('/relationships/user/:userId', async (req, res) => {
  try {
    const { userId } = req.params;
    const {
      relationship_type,
      min_strength,
      source_node_id,
      target_node_id,
      limit = 100,
      offset = 0
    } = req.query;

    let query = supabase
      .from('knowledge_relationships')
      .select(`
        *,
        source_node:knowledge_nodes!source_node_id(*),
        target_node:knowledge_nodes!target_node_id(*)
      `)
      .eq('user_id', userId);

    if (relationship_type) {
      query = query.eq('relationship_type', relationship_type);
    }

    if (min_strength) {
      query = query.gte('strength', parseFloat(min_strength));
    }

    if (source_node_id) {
      query = query.eq('source_node_id', source_node_id);
    }

    if (target_node_id) {
      query = query.eq('target_node_id', target_node_id);
    }

    query = query
      .order('strength', { ascending: false })
      .range(offset, offset + limit - 1);

    const { data, error } = await query;

    if (error) throw error;

    res.json({ relationships: data || [] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/knowledge/traverse/:userId/:nodeId - Traverse knowledge graph
router.get('/traverse/:userId/:nodeId', async (req, res) => {
  try {
    const { userId, nodeId } = req.params;
    const {
      relationship_types,
      max_depth = 3,
      min_strength = 0.3
    } = req.query;

    const relationshipTypesArray = relationship_types ?
      relationship_types.split(',') : null;

    const { data, error } = await supabase
      .rpc('traverse_knowledge_graph', {
        p_user_id: userId,
        p_start_node_id: nodeId,
        p_relationship_types: relationshipTypesArray,
        p_max_depth: parseInt(max_depth),
        p_min_strength: parseFloat(min_strength)
      });

    if (error) throw error;

    res.json({ graph: data || [] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/knowledge/path/:userId/:sourceId/:targetId - Find connection path
router.get('/path/:userId/:sourceId/:targetId', async (req, res) => {
  try {
    const { userId, sourceId, targetId } = req.params;
    const { max_depth = 5 } = req.query;

    const { data, error } = await supabase
      .rpc('find_connection_path', {
        p_user_id: userId,
        p_start_node_id: sourceId,
        p_end_node_id: targetId,
        p_max_depth: parseInt(max_depth)
      });

    if (error) throw error;

    res.json({ path: data || null });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/knowledge/similar/:userId - Find similar nodes
router.get('/similar/:userId', async (req, res) => {
  try {
    const { userId } = req.params;
    const {
      node_type,
      content_search,
      limit = 10
    } = req.query;

    const { data, error } = await supabase
      .rpc('find_similar_nodes', {
        p_user_id: userId,
        p_node_type: node_type,
        p_content_search: content_search,
        p_limit: parseInt(limit)
      });

    if (error) throw error;

    res.json({ similar_nodes: data || [] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/knowledge/analytics/:userId - Get knowledge graph analytics
router.get('/analytics/:userId', async (req, res) => {
  try {
    const { userId } = req.params;

    // Get nodes
    const { data: nodes, error: nodesError } = await supabase
      .from('knowledge_nodes')
      .select('*')
      .eq('user_id', userId);

    if (nodesError) throw nodesError;

    // Get relationships
    const { data: relationships, error: relsError } = await supabase
      .from('knowledge_relationships')
      .select('*')
      .eq('user_id', userId);

    if (relsError) throw relsError;

    const analytics = {
      nodes: {
        total: nodes.length,
        by_type: nodes.reduce((acc, n) => {
          acc[n.node_type] = (acc[n.node_type] || 0) + 1;
          return acc;
        }, {}),
        avg_weight: nodes.length ?
          nodes.reduce((sum, n) => sum + n.weight, 0) / nodes.length : 0,
        top_nodes: nodes
          .sort((a, b) => b.weight - a.weight)
          .slice(0, 10)
          .map(n => ({ id: n.id, weight: n.weight, type: n.node_type }))
      },
      relationships: {
        total: relationships.length,
        by_type: relationships.reduce((acc, r) => {
          acc[r.relationship_type] = (acc[r.relationship_type] || 0) + 1;
          return acc;
        }, {}),
        avg_strength: relationships.length ?
          relationships.reduce((sum, r) => sum + r.strength, 0) / relationships.length : 0,
        strongest: relationships
          .sort((a, b) => b.strength - a.strength)
          .slice(0, 10)
          .map(r => ({
            id: r.id,
            strength: r.strength,
            type: r.relationship_type,
            source: r.source_node_id,
            target: r.target_node_id
          }))
      },
      connectivity: {
        density: nodes.length > 1 ?
          relationships.length / (nodes.length * (nodes.length - 1)) : 0,
        isolated_nodes: nodes.filter(n =>
          !relationships.some(r =>
            r.source_node_id === n.id || r.target_node_id === n.id
          )
        ).length
      }
    };

    res.json({ analytics });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/knowledge/export/:userId - Export knowledge graph
router.get('/export/:userId', async (req, res) => {
  try {
    const { userId } = req.params;
    const { format = 'json' } = req.query;

    // Get all nodes and relationships
    const { data: nodes, error: nodesError } = await supabase
      .from('knowledge_nodes')
      .select('*')
      .eq('user_id', userId);

    if (nodesError) throw nodesError;

    const { data: relationships, error: relsError } = await supabase
      .from('knowledge_relationships')
      .select('*')
      .eq('user_id', userId);

    if (relsError) throw relsError;

    const exportData = {
      export_timestamp: new Date().toISOString(),
      user_id: userId,
      nodes: nodes || [],
      relationships: relationships || [],
      metadata: {
        total_nodes: nodes?.length || 0,
        total_relationships: relationships?.length || 0,
        node_types: [...new Set(nodes?.map(n => n.node_type) || [])],
        relationship_types: [...new Set(relationships?.map(r => r.relationship_type) || [])]
      }
    };

    if (format === 'cypher') {
      // Convert to Cypher queries for Neo4j import
      const cypherQueries = [];

      // Create nodes
      nodes?.forEach(node => {
        const props = JSON.stringify({
          id: node.id,
          weight: node.weight,
          content: node.content,
          created_at: node.created_at
        });
        cypherQueries.push(`CREATE (n:${node.node_type} ${props})`);
      });

      // Create relationships
      relationships?.forEach(rel => {
        cypherQueries.push(
          `MATCH (a {id: "${rel.source_node_id}"}), (b {id: "${rel.target_node_id}"}) ` +
          `CREATE (a)-[:${rel.relationship_type.toUpperCase()} {strength: ${rel.strength}}]->(b)`
        );
      });

      res.json({ cypher_queries: cypherQueries });
    } else {
      res.json(exportData);
    }
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: squire-backend/app/routers/knowledge.py
"""
Knowledge Graph routes
"""
from fastapi import APIRouter, HTTPException, Depends, Query
from typing import Optional, List
from uuid import UUID

from app.core.database import get_supabase, execute_rpc, execute_query, DatabaseError
from app.models.schemas import (
    KnowledgeNode,
    KnowledgeNodeCreate,
    KnowledgeNodeUpdate,
    KnowledgeRelationship,
    KnowledgeRelationshipCreate,
    NodeType,
    RelationshipType,
    SuccessResponse
)

router = APIRouter()


@router.post("/nodes", response_model=dict)
async def create_knowledge_node(
    node_data: KnowledgeNodeCreate,
    supabase=Depends(get_supabase)
):
    """Create/update knowledge node"""
    try:
        node_id = await execute_rpc(
            "upsert_knowledge_node",
            {
                "p_user_id": str(node_data.user_id),
                "p_node_type": node_data.node_type.value,
                "p_content": node_data.content,
                "p_weight": node_data.weight,
                "p_source_event_ids": [str(eid) for eid in node_data.source_event_ids],
                "p_metadata": node_data.metadata
            }
        )

        node = await execute_query(
            table="knowledge_nodes",
            operation="select",
            filters={"id": str(node_id)},
            single=True
        )

        return {
            "node": KnowledgeNode(**node),
            "node_id": node_id
        }
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/nodes/user/{user_id}", response_model=dict)
async def get_user_knowledge_nodes(
    user_id: UUID,
    node_type: Optional[NodeType] = Query(None),
    min_weight: Optional[float] = Query(None, ge=0.0),
    limit: int = Query(100, le=1000),
    offset: int = Query(0, ge=0),
    order_by: str = Query("weight"),
    order_direction: str = Query("desc"),
    supabase=Depends(get_supabase)
):
    """Get all knowledge nodes for user"""
    try:
        filters = {"user_id": str(user_id)}

        if node_type:
            filters["node_type"] = node_type.value

        nodes = await execute_query(
            table="knowledge_nodes",
            operation="select",
            filters=filters,
            order_by=order_by,
            ascending=(order_direction == "asc"),
            limit=limit,
            offset=offset
        )

        if nodes and min_weight is not None:
            nodes = [n for n in nodes if n.get("weight", 0) >= min_weight]

        return {"nodes": [KnowledgeNode(**node) for node in (nodes or [])]}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/relationships", response_model=dict)
async def create_knowledge_relationship(
    relationship_data: KnowledgeRelationshipCreate,
    supabase=Depends(get_supabase)
):
    """Create/update knowledge relationship"""
    try:
        relationship_id = await execute_rpc(
            "upsert_knowledge_relationship",
            {
                "p_user_id": str(relationship_data.user_id),
                "p_source_node_id": str(relationship_data.source_node_id),
                "p_target_node_id": str(relationship_data.target_node_id),
                "p_relationship_type": relationship_data.relationship_type.value,
                "p_strength": relationship_data.strength,
                "p_metadata": relationship_data.metadata
            }
        )

        relationship = await execute_query(
            table="knowledge_relationships",
            operation="select",
            filters={"id": str(relationship_id)},
            single=True
        )

        return {
            "relationship": KnowledgeRelationship(**relationship),
            "relationship_id": relationship_id
        }
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/traverse/{user_id}/{node_id}", response_model=dict)
async def traverse_knowledge_graph(
    user_id: UUID,
    node_id: UUID,
    relationship_types: Optional[str] = Query(None),
    max_depth: int = Query(3, ge=1, le=10),
    min_strength: float = Query(0.3, ge=0.0, le=1.0),
    supabase=Depends(get_supabase)
):
    """Traverse knowledge graph"""
    try:
        relationship_types_list = None
        if relationship_types:
            relationship_types_list = relationship_types.split(",")

        graph = await execute_rpc(
            "traverse_knowledge_graph",
            {
                "p_user_id": str(user_id),
                "p_start_node_id": str(node_id),
                "p_relationship_types": relationship_types_list,
                "p_max_depth": max_depth,
                "p_min_strength": min_strength
            }
        )

        return {"graph": graph or []}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/path/{user_id}/{source_id}/{target_id}", response_model=dict)
async def find_connection_path(
    user_id: UUID,
    source_id: UUID,
    target_id: UUID,
    max_depth: int = Query(5, ge=1, le=10),
    supabase=Depends(get_supabase)
):
    """Find connection path between nodes"""
    try:
        path = await execute_rpc(
            "find_connection_path",
            {
                "p_user_id": str(user_id),
                "p_start_node_id": str(source_id),
                "p_end_node_id": str(target_id),
                "p_max_depth": max_depth
            }
        )

        return {"path": path}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/similar/{user_id}", response_model=dict)
async def find_similar_nodes(
    user_id: UUID,
    node_type: Optional[NodeType] = Query(None),
    content_search: Optional[str] = Query(None),
    limit: int = Query(10, le=100),
    supabase=Depends(get_supabase)
):
    """Find similar nodes"""
    try:
        similar_nodes = await execute_rpc(
            "find_similar_nodes",
            {
                "p_user_id": str(user_id),
                "p_node_type": node_type.value if node_type else None,
                "p_content_search": content_search,
                "p_limit": limit
            }
        )

        return {"similar_nodes": similar_nodes or []}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/analytics/{user_id}", response_model=dict)
async def get_knowledge_analytics(
    user_id: UUID,
    supabase=Depends(get_supabase)
):
    """Get knowledge graph analytics"""
    try:
        nodes = await execute_query(
            table="knowledge_nodes",
            operation="select",
            filters={"user_id": str(user_id)}
        )

        relationships = await execute_query(
            table="knowledge_relationships",
            operation="select",
            filters={"user_id": str(user_id)}
        )

        analytics = {
            "nodes": {
                "total": len(nodes or []),
                "by_type": {},
                "avg_weight": 0,
                "top_nodes": []
            },
            "relationships": {
                "total": len(relationships or []),
                "by_type": {},
                "avg_strength": 0,
                "strongest": []
            },
            "connectivity": {
                "density": 0,
                "isolated_nodes": 0
            }
        }

        if nodes:
            # Calculate by type
            for node in nodes:
                node_type = node.get("node_type", "unknown")
                analytics["nodes"]["by_type"][node_type] = analytics["nodes"]["by_type"].get(node_type, 0) + 1

            # Calculate average weight
            weights = [n.get("weight", 0) for n in nodes]
            analytics["nodes"]["avg_weight"] = sum(weights) / len(weights)

            # Top nodes by weight
            sorted_nodes = sorted(nodes, key=lambda x: x.get("weight", 0), reverse=True)[:10]
            analytics["nodes"]["top_nodes"] = [
                {"id": n["id"], "weight": n.get("weight", 0), "type": n.get("node_type")}
                for n in sorted_nodes
            ]

        if relationships:
            # Calculate by type
            for rel in relationships:
                rel_type = rel.get("relationship_type", "unknown")
                analytics["relationships"]["by_type"][rel_type] = analytics["relationships"]["by_type"].get(rel_type, 0) + 1

            # Calculate average strength
            strengths = [r.get("strength", 0) for r in relationships]
            analytics["relationships"]["avg_strength"] = sum(strengths) / len(strengths)

            # Strongest relationships
            sorted_rels = sorted(relationships, key=lambda x: x.get("strength", 0), reverse=True)[:10]
            analytics["relationships"]["strongest"] = [
                {
                    "id": r["id"],
                    "strength": r.get("strength", 0),
                    "type": r.get("relationship_type"),
                    "source": r.get("source_node_id"),
                    "target": r.get("target_node_id")
                }
                for r in sorted_rels
            ]

        # Calculate connectivity
        if nodes and len(nodes) > 1:
            max_connections = len(nodes) * (len(nodes) - 1)
            analytics["connectivity"]["density"] = len(relationships or []) / max_connections

            # Find isolated nodes
            connected_nodes = set()
            for rel in (relationships or []):
                connected_nodes.add(rel.get("source_node_id"))
                connected_nodes.add(rel.get("target_node_id"))

            analytics["connectivity"]["isolated_nodes"] = len(nodes) - len(connected_nodes)

        return {"analytics": analytics}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))